provider/oracle/environ.go

```diff
@@ -0,0 +1,744 @@
+	}
+
+	tools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("Tools: %v", tools)
+	if err = args.InstanceConfig.SetTools(tools); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if err = instancecfg.FinishInstanceConfig(
+		args.InstanceConfig,
+		o.Config(),
+	); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// new cloud config template based on the series
+	cloudcfg, err := cloudinit.New(args.InstanceConfig.Series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
+
+	// compose userdata with the cloud config template
+	userData, err := providerinit.ComposeUserData(
+		args.InstanceConfig,
+		cloudcfg,
+		OracleRenderer{},
+	)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot make user data")
+	}
+
+	logger.Debugf("oracle user data: %d bytes", len(userData))
+
+	attributes := map[string]interface{}{
+		"userdata": string(userData),
+	}
+
+	hostname, err := o.namespace.Hostname(args.InstanceConfig.MachineId)
+	machineName := o.client.ComposeName(hostname)
+	imageName := o.client.ComposeName(imagelist)
+
+	tags := make([]string, 0, len(args.InstanceConfig.Tags)+1)
+	for k, v := range args.InstanceConfig.Tags {
+		if k == "" || v == "" {
+			continue
+		}
+		t := tagValue{k, v}
+		tags = append(tags, t.String())
+	}
+	tags = append(tags, machineName)
+
+	var apiPort int
+	if args.InstanceConfig.Controller != nil {
+		apiPort = args.InstanceConfig.Controller.Config.APIPort()
+	} else {
+		// All ports are the same so pick the first.
+		apiPort = args.InstanceConfig.APIInfo.Ports()[0]
+	}
+
+	// create a new seclists
+	secLists, err := o.firewall.CreateMachineSecLists(
+		args.InstanceConfig.MachineId, apiPort)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// create a  new netowrking card used for making the instance
+	// have a public address ip
+	networking := map[string]oci.Networker{
+		"eth0": oci.SharedNetwork{
+			Seclists: secLists,
+		},
+	}
+
+	// create the instance based on the instance params
+	instance, err := o.createInstance(o.client, oci.InstanceParams{
+		Instances: []oci.Instances{
+			{
+				Shape:       spec.InstanceType.Name,
+				Imagelist:   imageName,
+				Name:        machineName,
+				Label:       args.InstanceConfig.MachineAgentServiceName,
+				Hostname:    args.InstanceConfig.MachineAgentServiceName,
+				Tags:        tags,
+				Attributes:  attributes,
+				Reverse_dns: false,
+				Networking:  networking,
+			},
+		},
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	instance.arch = &spec.Image.Arch
+	instance.instType = &spec.InstanceType
+
+	machineId := instance.machine.Name
+	timeout := 10 * time.Minute
+	if err := instance.waitForMachineStatus(ociCommon.StateRunning, timeout); err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Infof("started instance %q", machineId)
+	//TODO: add config option for public IP allocation
```

Comment:
I thought about this a bit more. This is just a thought, and will not be addressed here, but I'm leaving a note here for posterity.

Currently the behavior is to either allocate a floating IP by default or not (based on provider config). I think the option of whether or not to allocate a public IP to a unit should be addressed by a separate juju command, and here's why:

Consider that most times the workload you are deploying does not need a public facing IP. Most services that get deployed to a cloud are comprised of a bunch of workers that service requests passed to them by a load balancer or a web server, or a database server that is consumed by other workloads you are deploying via juju in the same cloud. In this scenario, the only workload that needs a public IP is the load balancer. Assigning an IP when exposing an application may also not be optimal. On some (maybe all?) clouds, in order for the load balancer to be able to communicate with the worker servers, you need to expose the workers. They still don't need a public IP, but it's important to open up the provider firewall to allow access from the load balancer to the worker itself.

In this case, I think the most sane approach would be to do public IP management via a juju command. Something like:

```bash
juju public-ip add u/0
```

But, as I said, this is just a thought that might be worth investigating some other time :)